test/lit/passes/type-refining.wast

```diff
@@ -768,3 +768,70 @@
     )
   )
 )
+
+(module
+  (type $B (struct_subtype  $D))
+  ;; CHECK:      (type $G (struct_subtype  $D))
+  (type $G (struct_subtype  $D))
+
+  ;; CHECK:      (type $F (func_subtype func))
+  (type $F (func_subtype func))
+
+  ;; CHECK:      (type $I (struct_subtype (field (ref $G)) $C))
+  (type $I (struct_subtype (field (ref $H)) $C))
+
+ ;; CHECK:      (type $E (struct_subtype (field (ref $G)) $C))
+ (type $E (struct_subtype (field (ref $G)) $C))
+
+  ;; CHECK:      (type $C (struct_subtype (field (ref $G)) data))
+  (type $C (struct_subtype (field (ref $D)) data))
+
+  ;; CHECK:      (type $D (struct_subtype  data))
+  (type $D (struct_subtype  data))
+
+  (type $H (struct_subtype (field (ref $F)) $B))
+
+  ;; CHECK:      (func $func (type $F)
+  ;; CHECK-NEXT:  (drop
+  ;; CHECK-NEXT:   (block ;; (replaces something unreachable we can't emit)
+  ;; CHECK-NEXT:    (drop
+  ;; CHECK-NEXT:     (block
+  ;; CHECK-NEXT:      (drop
+  ;; CHECK-NEXT:       (ref.null $I)
+  ;; CHECK-NEXT:      )
+  ;; CHECK-NEXT:      (unreachable)
+  ;; CHECK-NEXT:     )
+  ;; CHECK-NEXT:    )
+  ;; CHECK-NEXT:   )
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT:  (drop
+  ;; CHECK-NEXT:   (struct.new $E
+  ;; CHECK-NEXT:    (struct.new_default $G)
+  ;; CHECK-NEXT:   )
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT: )
+  (func $func
+    (drop
+      ;; We never create an instance of $I, and we only have a get of its
+      ;; field. This optimization ignores the get (to not be limited by it).
+      ;; It will then optimize $I's field of $H (another struct for which we
+      ;; have no creation, and only a get) into $G, which is driven by the fact
```

Comment:
Why do we change the field to a completely unrelated type? This issue wouldn't happen if we didn't do that, right? Maybe we should be even more explicit and change the field type to `unreachable` or some other fresh type in cases where it can't possibly exist, then go fix up all the dead code (and make the type valid again) afterwards.

Reply:
We basically don't know what to do when we have zero constraints. We have to pick *some* type, so we pick the one used by our supertype, see details here: https://github.com/WebAssembly/binaryen/blob/4e8b85a8d3953234bbfa490cb1d4f5c32380fb3f/src/passes/TypeRefining.cpp#L134-L172 And the supertype might have picked up a type from another branch of the type tree, which is not a subtype of ours.

The key issue, though, is simple, I think: Our constraints are writes, and we ignore reads. So if we have only reads, we seem to have no constraints, and logically we can do "anything". That is the right way to model it, I think, as we've done. But we do need to fix up the case where "anything" happens to cause a validation error.

I don't see how to mark such types as a special type without altering the entire algorithm here, but maybe I'm missing something. The issue is that a type may have no reads or writes, but its parent or subtypes might. When we propagate information we have something to propagate from the parent or subtypes - again, they do have information - so it overrides us. So we wouldn't end up with the special type any more. A more complex encoding might handle that, I guess, but again, I think the key issue is simpler to think about as mentioned in the last paragraph.

Reply:
Aha, "We have to pick some type, so we pick the one used by our supertype... And the supertype might have picked up a type from another branch of the type tree, which is not a subtype of ours." makes it clear. Can you add that at the beginning of the comment here? It also might help to give the types names describing their role rather than just letters. Also, would it be possible to construct a test case with fewer types?

Reply:
Added.

I managed to remove one middle type, and another field I turned from a ref to an i32, while keeping the bug present, so it is simpler now. I also added names as you suggested, and a little diagram. Hopefully it's clearer!